src/main/java/dev/mccue/resolve/core/CacheHandler.java

```diff
@@ -0,0 +1,84 @@
+package dev.mccue.resolve.core;
+
+import java.io.FileOutputStream;
+import java.net.URL;
+import java.nio.channels.Channels;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Locale;
+
+public final class CacheHandler {
+    public enum OSLabel {
+        Linux, MacOS, Windows
+    }
+    private static OSLabel currentOS;
+    private static String CACHE_PATH;
+
+    public CacheHandler() {
+        /* This only handles the default cache path and will need to
+        *  be modified when new features are implemented (e.g. CLI --cache option)
+        */
+        checkOS();
+        switch (currentOS){
+            case Linux:
+                CACHE_PATH = System.getProperty("user.home") + "/.cache/jresolve";
+                break;
+            case MacOS:
```

Comment:
This exact bit of code, under the right concurrent conditions, could erroneously get to the `"Operating System could not be determined."` branch even if the OS is known.

Welcome to heck

https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/